### PR TITLE
Remove dependancy on legacy state for application bindings

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch removes the internal dependency on spaces in the legacy mongodb state regarding endpoint bindings. We do this by injecting the list of all available spaces for lookup, which needed for validating the bindings.

This change requires passing this new value into the EndpointBindings() method which itself requires many other function signatures to be changed in state, namely AddApplication(). This explains the huge number of modified files and lines of code.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

A semi-complex scenario regarding integration and CMR would be a good test:

```
# Bootstrap a controller using juju 3.4 and deploy influxdb:
juju bootstrap localhost ctrl-src1
juju add-model model-src1
juju deploy influxdb
juju offer influxdb:grafana-source  

# Bootstrap a second controller using this branch (3.5) and deploy grafana:
juju bootstrap localhost ctrl-src2
juju add-model model-src2
juju deploy grafana

# Integration between grafana and influxdb should work:
juju integrate grafana ctrl-src1:admin/model-src1.influxdb

juju status --relations
Model       Controller  Cloud/Region         Version      SLA          Timestamp
model-src2  ctrl-src2   localhost/localhost  3.5-beta1.1  unsupported  12:09:18+01:00

SAAS      Status  Store      URL
influxdb  active  ctrl-src1  admin/model-src1.influxdb

App      Version  Status  Scale  Charm    Channel        Rev  Exposed  Message
grafana           active      1  grafana  latest/stable   69  no       Ready

Unit        Workload  Agent  Machine  Public address  Ports     Message
grafana/0*  active    idle   0        10.170.236.239  3000/tcp  Ready

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.170.236.239  juju-3f7181-0  ubuntu@20.04      Running

Integration provider     Requirer                Interface       Type     Message
influxdb:grafana-source  grafana:grafana-source  grafana-source  regular
```

## Links

**Jira card:** JUJU-5107

### CODE REVIEW

File: state/endpoint_bindings.go [hunk 1]
```diff
@@ -464,43 +452,30 @@ func DefaultEndpointBindingsForCharm(_ EndpointBinding, charmMeta *charm.Meta) (
...
 
 // NewBindings returns a bindings guaranteed to be in space id format.
-func NewBindings(st EndpointBinding, givenMap map[string]string) (*Bindings, error) {
-	// namesErr and idError are only problems if both are not nil.
-	spaceInfos, err := st.AllSpaceInfos()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-
+func NewBindings(allSpaces network.SpaceInfos, givenMap map[string]string) (*Bindings, error) {
```
Comment (nvinuesa) [inline state/endpoint_bindings.go:463]: Main change right here

File: domain/network/service/space.go [hunk 2]
```diff
@@ -68,6 +68,11 @@ func (s *SpaceService) GetAllSpaces(ctx context.Context) (network.SpaceInfos, er
 	return s.st.GetAllSpaces(ctx)
 }
 
+// UpdateSpace updates the space identified by the passed uuid.
+func (s *SpaceService) UpdateSpace(ctx context.Context, uuid string, name string) error {
+	return s.st.UpdateSpace(ctx, uuid, name)
```
Comment (manadart) [inline domain/network/service/space.go:73]: Trace this like you did for `GetAllSubnets`

File: apiserver/facades/client/subnets/cache.go [hunk 3]
```diff
@@ -18,34 +17,6 @@ import (
 	"github.com/juju/juju/rpc/params"
 )
 
-// addSubnetsCache holds cached lists of spaces, zones, and subnets, used for
-// fast lookups while adding subnets.
-type addSubnetsCache struct {
```
Comment (manadart) [inline apiserver/facades/client/subnets/cache.go:23]: I'd like to remove this notion of cache altogether.

File: apiserver/common/networkingcommon/types.go [hunk 4]
```diff
@@ -7,86 +7,12 @@ import (
...
-// backing store.
-//
-// TODO(dimitern): Replace state.SubnetInfo with this and remove
-// BackingSubnetInfo, once the rest of state backing methods and the
-// following pre-reqs are done:
-//   - Subnets need a reference count to calculate Status.
-//   - ensure EC2 and MAAS providers accept empty IDs as Subnets() args
-//     and return all subnets, including the AvailabilityZones (for EC2;
-//     empty for MAAS as zones are orthogonal to networks).
-type BackingSubnetInfo struct {
```
Comment (manadart) [inline apiserver/common/networkingcommon/types.go:44]: Glad to be getting rid of these.

File: apiserver/facades/agent/machine/package_test.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2013 Canonical Ltd.
+// Copyright 2024444 Canonical Ltd.
```
Comment (manadart) [inline apiserver/facades/agent/machine/package_test.go:1]: Whoopsie.

File: state/endpoint_bindings.go [hunk 6]
```diff
@@ -84,13 +84,13 @@ func (b bindingsMap) GetBSON() (interface{}, error) {
 func (b *Bindings) Merge(mergeWith map[string]string, meta *charm.Meta) (bool, error) {
 	// Verify the bindings to be merged, and ensure we're merging with
 	// space ids.
-	merge, err := NewBindings(b.st, mergeWith)
+	merge, err := NewBindings(b.allSpaces, mergeWith)
```
Comment (manadart) [inline state/endpoint_bindings.go:87]: I think we might have saved a lot of churn by removing `newBindingsFromNames` (it was for backwards compatibility), and just assuming that supplied binding IDs were valid.

**Review by manadart — COMMENTED** [review]
We might need to talk about this.

**Pull request closed without merging**